### PR TITLE
Fix context menu focus for Edge downloads

### LINKED ISSUE #14916: Microsoft edge: unable to redirect focus to context menu in downloads window
### Steps to reproduce:
1. open Microsoft Edge;
2.Try to download something;
3.press control+j to open downloas window;
4.press Applications key or shift+f10 to invoke context menu.
### Actual behavior:
Context menu opens, but NVDA focus is not redirected there and arrow keys does not do anything.
### Expected behavior:
Context menues should be read as in previous msedge versions.
#### NVDA installed/portable/running from source:
Installed
#### NVDA version:
alpha-28193,d1ce3bc9 (2023.2.0.28193)
#### Windows version:
Windows 10 22H2 (AMD64) build 19045.2728
#### Name and version of other software in use when reproducing the issue:
Microsoft Edge Version 113.0.1774.35 (Official build) (64-bit) Version 113.0.1774.35 (Official build) (64-bit)
#### Does the issue still occur after restarting your computer?
Yes
#### Have you tried any other versions of NVDA? If so, please report their behaviors.
No
#### If NVDA add-ons are disabled, is your problem still occurring?
Yes
#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
Yes

### PR DESCRIPTION
**Pull request by LeonarddeR**
Note, created against beta because this issue can be pretty annoying and it is a relatively small fix with a major positive impact for Edge users.

### Link to issue number:
Fixes #14916 

### Summary of the issue:
When opening the context menu from the Edge downloads window, NVDA doesn't reflect this.

### Description of user facing changes
Context menu is read again.

### Description of development approach
1. It looks like the context menu items aren't direct children of the foreground window, therefore we compare the root of the foreground and the root of the focus window instead.
2. Fixed a typo in a UIAHandler debug string
3. Ensured that `UIAHandler.getNearestWindowHandle` returns when a cached window handle couldn't be fetched from an UIAElement due to a COMError.

### Testing strategy:
Tested that the context menu of a download item in Edge reads again.

### Known issues with pull request:
The first item in the menu is read twice. I've tried to fix this without success. It looks like our duplicate focus event logic doesn't kick in properly. Note that Edge firing duplicate events is a bug on their end anyway.

### Change log entries:
Bug fixes
- NVDA is once again able to read the context menu of downloads in Microsoft Edge. (#14916)

### Code Review Checklist:

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] Security precautions taken.

### CODE REVIEW

**Comment by Carlos-EstebanM** [comment]
Hello all.
I tested the compilation of this PR in a virtual machine with windows 11 and Microsoft Edge 115.0.1901.203 stable. NVDA can’t read the context menu, but there is not an error in the log, check with information and debug level. I also install and test with Microsoft Edge beta 116.0.1938.43, with the same result. I test with a clean and install copy on the virtual machine, and a friend test in his physical machine with the portable and temporally copy of this compilation, same versions of Windows and Microsoft Edge.
Regards.

**Comment by cary-rowen** [comment]
Hi,

For me that fix worked.
MSEdge-115.0.1901.203 
Windows 10 22H2 (AMD64) build 19045.3324

**Comment by leonardder** [comment]
@Carlos-EstebanM Thanks for reporting this. Here it works as expected. Could you please ensure the following?

1. It is indeed the right build you're testing, namely [this one](https://ci.appveyor.com/api/buildjobs/ymqrrc67jvsk9s41/artifacts/output%2Fnvda_snapshot_pr15300-28888%2C5da41e4e.exe)
2. You're using the right steps to reproduce, i.e.
    1. Open Edge
    2. Press CTRL+J to open the downloads window
    3. To ensure proper navigation, disable browse mode if it is enabled (NVDA+SPace)
    4. Focus a download (either completed or running)
    5. Press shift+f10 or the applications key.

If this doesn't work, please provide a log as follows:

1. Set NVDA log level to debug
2. In the advanced settings category, debug logging grouping, enable the UIA category
3. Focus a download in Edge
4. Press control+shift+nvda+f1 to start log collection
5. Press either applications key or shift+f10 to open the context menu
6. Press control+shift+nvda+f1 again to copy the relevant log snippet to the clip board
7. Paste the snippet in this issue

**Comment by XLTechie** [comment]
@leonardder I also don't seem to be having success with this. That said, I'm not an Edge user, normally, so perhaps I'm doing it wrong.
Pressing Ctrl+j, Going into focus mode, tab to the file name, press context key. Nothing happens, can't shift+tab, NVDA is silent.
Edge 115.0.1901.203
Windows 10.0.22621.1 (11 Home 22H2)
Running a temporary copy of this PR

[Log Attached](https://github.com/nvaccess/nvda/files/12357878/log-for-edge-downloads.txt)

**Comment by leonardder** [comment]
It looks like in your case the window class is Chrome_WidgetWin_1, not Chrome_WidgetWin_2. I changed the code to support all incarnations of Chrome_WidgetWin_*

File: source/UIAHandler/__init__.py [hunk 1]
```diff
@@ -1237,7 +1237,7 @@ def getNearestWindowHandle(self, UIAElement):
 		try:
 			window = new.cachedNativeWindowHandle
 		except COMError:
-			window = None
+			return None
```
Comment (leonardder) [inline source/UIAHandler/__init__.py:1240]: Note, this fixes two issues:

1. NVDA trying to call `getWindowHandleDebugString` on a None window handle. This was discovered while debugging the issue with Edge.
2. Saving a None _nearestWindowHandle on an UIA element

**Comment by rperez030** [comment]
It works for me with the latest commit. like @Carlos, I also have an instance with the 'Chrome_WidgetWin_1' class. It also fixed the Folder menu in the add or edit favorite dialog as I expected. So far so good. Thanks @leonardder for such a quick turnaround.

**Comment by cary-rowen** [comment]
Hi,
I retested the latest commit and it still works.
Thanks

**Comment by seanbudd** [comment]
Apologies but this is much too late for a beta at this point. Unless another serious regression is found, we don't plan on releasing another beta.

**Comment by cary-rowen** [comment]
But I think this bug seriously affects the experience of NVDA users. This is a bug specific to NVDA and Edge.
Although I'm not an Edge user, I'd like to see this fix in the latest stable release.

**Comment by wmhn1872265132** [comment]
Since Microsoft Edge is the pre-installed browser of Windows 10 and 11, this problem may affect many users, and the scenarios involved are also critical. I really hope to consider fixing it in time in this version

File: source/UIAHandler/__init__.py (hunk 1 above)
Comment (seanbudd) [reply source/UIAHandler/__init__.py:1240]: can we a log something at debugWarning here?

File: source/eventHandler.py [hunk 2]
```diff
@@ -499,11 +499,18 @@ def shouldAcceptEvent(eventName, windowHandle=None):
 
 	fg = winUser.getForegroundWindow()
 	fgClassName=winUser.getClassName(fg)
-	if wClass == "NetUIHWND" and fgClassName in ("Net UI Tool Window Layered","Net UI Tool Window"):
+	if (
 		# #5504: In Office >= 2013 with the ribbon showing only tabs,
 		# when a tab is expanded, the window we get from the focus object is incorrect.
 		# This window isn't beneath the foreground window,
-		# so our foreground application checks fail.
```
Comment (seanbudd) [inline source/eventHandler.py:506]: I think this should be kept, or the previous line should get the comma replaced with a period

**Comment by XLTechie** [comment]
@seanbudd Given the volume of complaints and users effected, and given how long 
this has been an outstanding issue (long before the first beta): as you're not 
inclined to do another beta cycle to include this, would a 2023.2.1 be possible?

Though I would think a third beta would be easier, and less confusing to the 
users.

For myself I don't care, as I don't use Edge more than once a quarter, but users 
really want this fix, if mailing list traffic and number of issues about Edge 
context menus are any measure.

File: source/eventHandler.py (hunk 2 above)
Comment (leonardder) [reply source/eventHandler.py:506]: Note that I kept that line, but moved it further below the if check.

**Comment by seanbudd** [comment]
It seems that a bad merge to master occurred - there are a lot of unrelated changes in the dif

**Comment by seanbudd** [comment]
ah - I see it was a merge from beta - can this be rebased on master instead?

**Comment by leonardder** [comment]
> ah - I see it was a merge from beta - can this be rebased on master instead?

I don't think there's consensus yet about whether this has to go to 2023.2, 2023.3 or a potential 2023.2.1 😉

**Comment by cary-rowen** [comment]
Hi @seanbudd 
1. A user raised [this topic](https://nvda.groups.io/g/nvda/topic/microsoft_edge_114_issues/100073783) on Groups.io early on and asked if there is an issue about this bug being opened on github.
2. Even on github, there are multiple Issues related to this bug that have been created successively.
Such as: #14916, #15293, #15153, #15055, #14922

Based on the above, I think community users are suffering from this bug, can you reconsider giving this fix to beta branch?

Thanks

**Comment by Nikita34196** [comment]
Hello. I personally use nvda and edge. I am one of those suffering from this problem. I would like this to be released as soon as possible. I have many friends who use microsoft edge and this problem is quite annoying.

**Comment by rperez030** [comment]
> Apologies but this is much too late for a beta at this point. Unless another serious regression is found, we don't plan on releasing another beta.

In my opinion, a bug like this even deserves a hotfix. It affects multiple areas in the default browser. It can fully block at least 1 use case, in this case allowing the download of unsigned executables that are blocked by Windows defender. Users shouldn't wait an entire quarter or be forced to use development builds or install an add-on to get around an issue that blocks functionality in the default browser.

**Review by seanbudd — APPROVED** [review]

**Pull request merged**